### PR TITLE
Enable ZeRO tests for CI, fix to/half function calls for LightningDistributedWrapper

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Enables ZeRO tests for the CI.

I also noticed that the device types were not correct, as we do not call the lightning module device hooks. This adds the correct functionality by using the device type mixins as we do for the lightning module suggested by @awaelchli 

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by Borda — APPROVED** [review]
add chlog?

**Comment by SeanNaren** [comment]
Will do, seems like it doesnt work out the box, trying to resolve

File: pytorch_lightning/utilities/device_dtype_mixin.py [hunk 1]
```diff
@@ -132,7 +132,8 @@ def cuda(self, device: Optional[int] = None) -> Module:
         Returns:
             Module: self
         """
-        self.__update_properties(device=torch.device('cuda', index=device))
+        property_device = device if isinstance(device, torch.device) else torch.device('cuda', index=device)
+        self.__update_properties(device=property_device)
```
Comment (SeanNaren) [inline pytorch_lightning/utilities/device_dtype_mixin.py:136]: Let's make a test for this in particular, DP was passing in torch devices which using the standard `cuda` function work fine, but for our device mixins, they were unsupported which raised an error once we are not moving devices correctly.

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/utilities/device_dtype_mixin.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/utilities/device_dtype_mixin.py:136]: We have tests/utilities/test_dtype_device_mixin.py, we can add it there.

**Review by awaelchli — APPROVED** [review]

File: tests/plugins/test_deepspeed_plugin.py [hunk 2]
```diff
@@ -8,11 +8,52 @@
...
+    module = LightningDeepSpeedModule(model, precision=16)
+
+    module.cuda().half()
+    assert module.dtype == torch.half
+    assert model.dtype == torch.half
+
+    x = torch.randn((1, 32), dtype=torch.float).cuda()
+    out = module(x)
+
+    assert out.dtype == torch.half
```
Comment (tchaton) [inline tests/plugins/test_deepspeed_plugin.py:50]: I am really suprised we didn't add this kind of test before.

**Review by tchaton — APPROVED** [review]
LGTM !

**Pull request merged**